tools/procps-3.1.5-LTP/ps/ps.1
.\" Man page for ps.
.\" Quick hack conversion by Albert Cahalan, 1998.
.\" Licensed under version 2 of the Gnu General Public License.
.\"
.\" This man page is a horrid hack because *roff sucks.
.\" The whole system is way obsolete. The internal header
.\" stuff must die, and will when I figure out how to kill it.
.\" I've already killed the wasteful left margin and screwy
.\" old perfect justification. Gross! You'd think someone
.\" invented this crap in 1973. Oh yeah, they did. Sorry.
.\"
.TH PS 1 "July 5, 1998" "Linux" "Linux User's Manual"
.SH \fRNAME\fR
ps \- report process status
.ad r
.na
.ss 12 0
.in 0
.nh
.nf

SYNOPSIS
ps [options]


DESCRIPTION
ps gives a snapshot of the current processes. If you want
a repetitive update of this status, use top. This man
page documents the /proc-based version of ps, or tries to.


COMMAND-LINE OPTIONS

This version of ps accepts several kinds of options.

Unix options may be grouped and must be preceeded by a dash.
BSD options may be grouped and must not be used with a dash.
Gnu long options are preceeded by two dashes.

Options of different types may be freely mixed.

Set the I_WANT_A_BROKEN_PS environment variable to force BSD syntax even
when options are preceeded by a dash. The PS_PERSONALITY environment
variable (described below) provides more detailed control of ps behavior.

SIMPLE PROCESS SELECTION
-A           select all processes
-N           negate selection
-a           select all with a tty except session leaders
-d           select all, but omit session leaders
-e           select all processes
T            select all processes on this terminal
a            select all processes on a terminal, including those of other users
g            really all, even group leaders (does nothing w/o SunOS settings)
r            restrict output to running processes
x            select processes without controlling ttys
--deselect   negate selection

PROCESS SELECTION BY LIST
-C           select by command name
-G           select by RGID (supports names)
-U           select by RUID (supports names)
-g           select by session leader OR by group name
-p           select by PID
-s           select processes belonging to the sessions given
-t           select by tty
-u           select by effective user ID (supports names)
U            select processes for specified users
p            select by process ID
t            select by tty
--Group      select by real group name or ID
--User       select by real user name or ID
--group      select by effective group name or ID
--pid        select by process ID
--sid        select by session ID
--tty        select by terminal
--user       select by effective user name or ID
-123         implied --sid
123          implied --pid

OUTPUT FORMAT CONTROL
-O           is preloaded "-o"
-F           extra full format
-c           different scheduler info for -l option
-f           does full listing
-j           jobs format
-l           long format
-o           user-defined format
-y           do not show flags; show rss in place of addr
O            is preloaded "o" (overloaded)
X            old Linux i386 register format
j            job control format
l            display long format
o            specify user-defined format
s            display signal format
u            display user-oriented format
v            display virtual memory format
--format     user-defined format
--context    (SELinux only) Display security context format; implies --SID
--secsid     (SELinux only) Display Security ID (SID)
--SID	     (SELinux only) Display Security ID (SID)

OUTPUT MODIFIERS
-H           show process hierarchy (forest)
-m           show threads
-n           set namelist file
-w           wide output
C            use raw CPU time for %CPU instead of decaying average
N            specify namelist file
O            sorting order (overloaded)
S            include some dead child process data (as a sum with the parent)
c            true command name
e            show environment after the command
f            ASCII-art process hierarchy (forest)
h            no header (or, one header per screen in the BSD personality)
m            all threads
n            numeric output for WCHAN and USER
w            wide output
--cols       set screen width
--columns    set screen width
--cumulative include some dead child process data (as a sum with the parent)
--forest     ASCII art process tree
--headers    repeat header lines, one per page of output
--no-headers print no header line at all
--lines      set screen height
--rows       set screen height
--sort       specify sorting order
--width      set screen width

INFORMATION
-V          print version
L           list all format specifiers
V           show version info
--help      print help message
--info      print debugging info
--version   print version

OBSOLETE
A           increases the argument space (DecUnix)
M           use alternate core (try -n or N instead)
W           get swap info from ... not /dev/drum (try -n or N instead)
k           use /vmcore as c-dumpfile (try -n or N instead)



NOTES

The "-g" option can select by session leader OR by group name.
Selection by session leader is specified by many standards,
but selection by group is the logical behavior that several other
operating systems use. This ps will select by session leader when
the list is completely numeric (as sessions are). Group ID numbers
will work only when some group names are also specified.

The "m" option should not be used. Use "-m" or "-o" with a list.
("m" displays memory info, shows threads, or sorts by memory use)

The "h" option is problematic.  Standard BSD ps uses the option to
print a header on each page of output, but older Linux ps uses the option
to totally disable the header.  This version of ps follows the Linux
usage of not printing the header unless the BSD personality has been
selected, in which case it prints a header on each page of output.
Regardless of the current personality, you can use the long options
--headers and --no-headers to enable printing headers each page and
disable headers entirely, respectively.

Terminals (ttys, or screens for text output) can be specified in several
forms: /dev/ttyS1, ttyS1, S1. Obsolete "ps t" (your own terminal) and
"ps t?" (processes without a terminal) syntax is supported, but modern
options ("T", "-t" with list, "x", "t" with list) should be used instead.

The BSD "O" option can act like "-O" (user-defined output format with
some common fields predefined) or can be used to specify sort order.
Heuristics are used to determine the behavior of this option. To ensure
that the desired behavior is obtained, specify the other option (sorting
or formatting) in some other way.

For sorting, BSD "O" option syntax is O[+|-]k1[,[+|-]k2[,...]]
Order the process listing according to the multilevel sort specified by
the sequence of short keys from SORT KEYS, k1, k2, ... The `+' is quite
optional, merely re-iterating the default direction on a key. `-' reverses
direction only on the key it precedes. The O option must be the last option
in a single command argument, but specifications in successive arguments are
catenated.

Gnu sorting syntax is --sortX[+|-]key[,[+|-]key[,...]]
Choose a multi-letter key from the SORT KEYS section. X may be any
convenient separator character. To be GNU-ish use `='. The `+' is really
optional since default direction is increasing numerical or lexicographic
order. For example, ps jax --sort=uid,-ppid,+pid

This ps works by reading the virtual files in /proc. This ps does not
need to be suid kmem or have any privileges to run. Do not give this ps
any special permissions.

This ps needs access to a namelist file for proper WCHAN display.
The namelist file must match the current Linux kernel exactly for
correct output.

To produce the WCHAN field, ps needs to read the System.map file created
when the kernel is compiled. The search path is:

$PS_SYSTEM_MAP
/boot/System.map-`uname -r`
/boot/System.map
/lib/modules/`uname -r`/System.map
/usr/src/linux/System.map

The member used_math of task_struct is not shown, since crt0.s checks
to see if math is present. This causes the math flag to be set for all
processes, and so it is worthless. (Somebody fix libc or the kernel please)

Programs swapped out to disk will be shown without command line arguments,
and unless the c option is given, in brackets.

%CPU shows the cputime/realtime percentage. It will not add up to 100%
unless you are lucky. It is time used divided by the time the process has
been running.

The SIZE and RSS fields don't count the page tables and the task_struct of a
proc; this is at least 12k of memory that is always resident. SIZE is the
virtual size of the proc (code+data+stack).

Processes marked <defunct> are dead processes (so-called "zombies") that
remain because their parent has not destroyed them properly. These processes
will be destroyed by init(8) if the parent process exits.


PROCESS FLAGS

FORKNOEXEC   1    forked but didn't exec
SUPERPRIV    2    used super-user privileges
DUMPCORE     4    dumped core


PROCESS STATE CODES

D uninterruptible sleep (usually IO)
R runnable (on run queue)
S sleeping
T traced or stopped
W paging
X dead
Z a defunct ("zombie") process

For BSD formats and when the "stat" keyword is used, additional
letters may be displayed:

W has no resident pages
< high-priority process
N low-priority task
L has pages locked into memory (for real-time and custom IO)


SORT KEYS

Note that the values used in sorting are the internal values ps uses and not
the `cooked' values used in some of the output format fields. Pipe ps
output into the sort(1) command if you want to sort the cooked values.

KEY LONG       DESCRIPTION
c   cmd        simple name of executable
C   cmdline    full command line
f   flags      flags as in long format F field
g   pgrp       process group ID
G   tpgid      controlling tty process group ID
j   cutime     cumulative user time
J   cstime     cumulative system time
k   utime      user time
K   stime      system time
m   min_flt    number of minor page faults
M   maj_flt    number of major page faults
n   cmin_flt   cumulative minor page faults
N   cmaj_flt   cumulative major page faults
o   session    session ID
p   pid        process ID
P   ppid       parent process ID
r   rss        resident set size
R   resident   resident pages
s   size       memory size in kilobytes
S   share      amount of shared pages
t   tty        the minor device number of tty
T   start_time time process was started
U   uid        user ID number
u   user       user name
v   vsize      total VM size in kB
y   priority   kernel scheduling priority


AIX FORMAT DESCRIPTORS

This ps supports AIX format descriptors, which work somewhat like the
formatting codes of printf(1) and printf(3). For example, the normal
default output can be produced with this:   ps -eo "%p %y %x %c"

CODE  NORMAL    HEADER
%C    pcpu      %CPU
%G    group     GROUP
%P    ppid      PPID
%U    user      USER
%a    args      COMMAND
%c    comm      COMMAND
%g    rgroup    RGROUP
%n    nice      NI
%p    pid       PID
%r    pgid      PGID
%t    etime     ELAPSED
%u    ruser     RUSER
%x    time      TIME
%y    tty       TTY
%z    vsz       VSZ


STANDARD FORMAT SPECIFIERS

These may be used to control both output format and sorting.
For example:  ps -eo pid,user,args --sort user

CODE         HEADER
%cpu         %CPU    
%mem         %MEM    
alarm        ALARM   
args         COMMAND 
blocked      BLOCKED 
bsdstart     START   
bsdtime      TIME    
c            C       
caught       CAUGHT  
cmd          CMD     
comm         COMMAND 
command      COMMAND 
context	     CONTEXT
cputime      TIME    
drs          DRS     
dsiz         DSIZ    
egid         EGID    
egroup       EGROUP  
eip          EIP     
esp          ESP     
etime        ELAPSED 
euid         EUID    
euser        EUSER   
f            F       
fgid         FGID    
fgroup       FGROUP  
flag         F       
flags        F       
fname        COMMAND 
fsgid        FSGID   
fsgroup      FSGROUP 
fsuid        FSUID   
fsuser       FSUSER  
fuid         FUID    
fuser        FUSER   
gid          GID     
group        GROUP   
ignored      IGNORED 
intpri       PRI     
lim          LIM     
longtname    TTY     
lstart       STARTED 
m_drs        DRS     
m_trs        TRS     
maj_flt      MAJFL   
majflt       MAJFLT  
min_flt      MINFL   
minflt       MINFLT  
ni           NI      
nice         NI      
nwchan       WCHAN   
opri         PRI     
pagein       PAGEIN  
pcpu         %CPU    
pending      PENDING 
pgid         PGID    
pgrp         PGRP    
pid          PID     
pmem         %MEM    
ppid         PPID    
pri          PRI     
priority     PRI     
rgid         RGID    
rgroup       RGROUP  
rss          RSS     
rssize       RSS     
rsz          RSZ     
ruid         RUID    
ruser        RUSER   
s            S       
secsid       SECSID
sess         SESS    
session      SESS    
sgi_p        P       
sgi_rss      RSS     
sgid         SGID    
sgroup       SGROUP  
sid          SID     
sig          PENDING 
sig_block    BLOCKED 
sig_catch    CATCHED 
sig_ignore   IGNORED 
sig_pend     SIGNAL  
sigcatch     CAUGHT  
sigignore    IGNORED 
sigmask      BLOCKED 
stackp       STACKP  
start        STARTED 
start_stack  STACKP  
start_time   START   
stat         STAT    
state        S       
stime        STIME   
suid         SUID    
suser        SUSER   
svgid        SVGID   
svgroup      SVGROUP 
svuid        SVUID   
svuser       SVUSER  
sz           SZ      
time         TIME    
timeout      TMOUT   
tmout        TMOUT   
tname        TTY     
tpgid        TPGID   
trs          TRS     
trss         TRSS    
tsiz         TSIZ    
tt           TT      
tty          TT      
tty4         TTY     
tty8         TTY     
ucmd         CMD     
ucomm        COMMAND 
uid          UID     
uid_hack     UID     
uname        USER    
user         USER    
vsize        VSZ     
vsz          VSZ     
wchan        WCHAN   




ENVIRONMENT VARIABLES
The following environment variables could affect ps:
    COLUMNS             Override default display width.
    LINES               Override default display height.
    PS_PERSONALITY      Set to one of posix,old,linux,bsd,sun,digital...
    CMD_ENV             Set to one of posix,old,linux,bsd,sun,digital...
    I_WANT_A_BROKEN_PS  Force obsolete command line interpretation.
    LC_TIME             Date format.
    PS_COLORS           Not currently supported.
    PS_FORMAT           Default output format override.
    PS_SYSMAP           Default namelist (System.map) location.
    PS_SYSTEM_MAP       Default namelist (System.map) location.
    POSIXLY_CORRECT     Don't find excuses to ignore bad "features".
    UNIX95              Don't find excuses to ignore bad "features".
    _XPG                Cancel CMD_ENV=irix non-standard behavior.

In general, it is a bad idea to set these variables. The one exception
is CMD_ENV or PS_PERSONALITY, which could be set to Linux for normal
systems. Without that setting, ps follows the useless and bad parts
of the Unix98 standard.


PERSONALITY
    390      like the S/390 OpenEdition ps
    aix      like AIX ps
    bsd      like FreeBSD ps (totally non-standard)
    compaq   like Digital Unix ps
    debian   like the old Debian ps
    digital  like Digital Unix ps
    gnu      like the old Debian ps
    hp       like HP-UX ps
    hpux     like HP-UX ps
    irix     like Irix ps
    linux    ***** RECOMMENDED *****
    old      like the original Linux ps (totally non-standard)
    posix    standard
    sco      like SCO ps
    sgi      like Irix ps
    sun      like SunOS 4 ps (totally non-standard)
    sunos    like SunOS 4 ps (totally non-standard)
    sysv     standard
    unix     standard
    unix95   standard
    unix98   standard


EXAMPLES
To see every process on the system using standard syntax:
    ps -e
To see every process on the system using BSD syntax:
    ps ax

ChangeLog
LTP-20020307
------------
o	fixed IDcheck.sh to prompt for IDs only when missing (Jay Huie)
o	added time duration option to pan (Jay Huie)
o	fixed bug [ #516577 ] ftruncate03 test (Jay Huie)
o	added 4 more cases to getcwd02 test (Jay Huie)
o	added time-schedule tool by Richard Gooch under sched test
	section (Manoj Iyer)
o	added trace_sched tool under sched test section (Manoj Iyer)
o	fixed bug [ #511427 ] pread02 test (Paul Larson)
o	fixed bug [ #523055 ] sched_getscheduler test (Paul Larson)
o	fixed bug [ #523137 ] sched_setscheduler02 test (Paul Larson)
o	fixed bug [ #525688 ] sendfile01 (Paul Larson)
o	added sigpending02 test to syscalls section (Paul Larson)
o       created a KNOWN-FAILURES file to document known 
        test failures (Robbie Williamson)
o	added sem02 test to ipc section (Robbie Williamson)
o	added SCTP section to network tests (Robbie Williamson)
o	fixed bug [ #505515 ] perf_lan6 test (Robbie Williamson)
o	added disktest by Brent Yardley to io section (Robbie Williamson)
o	closed the following bugs:
	491285, 491286, 506536, 514408 (Paul Larson and Robbie Williamson)


LTP-20020207
------------
o       added support for cross-compiling (Todd Inglett)
o       added LKML's cache_leak testcase to ltp/scratch (Nate Straz)
o       added IPv6 support (Robbie Williamson)
o       added "gethost" to /tools (Robbie Williamson)
o	fixed the race conditions in the float tests and removed 
	the sleeps (Robbie Williamson)
o       enabled non-root make authority (Paul Larson)
o       separated compilation into "make" and "make install" (Paul Larson)
o       added ipc_stress test (Manoj Iyer)
o       added pthreads_stress test (Manoj Iyer)
o       made changes to support architecture independence (Manoj Iyer & Paul Larson)
o	closed the following bugs: 
	504960, 505108, 504613, 504616, 491283, 506689, 508055, 506692, 508074
	491289, 506662, 511383, 511391, 511427, 511494, 504649, 514050, 512953
	(Manoj Iyer, Paul Larson, and Robbie Williamson)	


LTP-20020108
------------
o	fixed IDcheck.sh to work with bash1 (Nate Straz)
o	updated menu script (Robbie Williamson)
o	reset errno to 0 at the beginning of TEST macro (Paul Larson)
o	added floating point stress tests (Bull/IBM)
o	added fsx-linux NFS tests (Apple)	
o	define GLIBC_SIGACTION_BUG for sigaction02 (Robbie Williamson)
o	removed bogus testcase from dup201 test (Paul Larson)
o	fixed recvfrom01 case 5 (Wayne Boyer)
o	workaround for smp issue with waitpid12 (Manfred Spraul)


LTP-20011206
------------
o	necessary users/groups can now be created on installation with 
	user's permission
o	added a simple menu-based interface for running the LTP testsuite
o	fixed negative duration in pan output when -l isn't used
o	new set of tests under fs-maim
o	fixed race condition in nfslock01
o	ar01, ld01, ldd01, nm01, objdump01, and size01 fixed when multiple 
	copies are run simultaneously
o	workaround for SIGTTOU hang in ioctl02
o	shmget03 uses IPC_PRIVATE to make it safe when running multiple copies
o	compiler warnings and other minor errors fixed in many tests

LTP-20011107
------------
o	Many improvements to mtest05 and mtest06 tests
o	new test nfsstress
o	included ver_linux in LTP and made it run at the end of test scripts
o	check for necessary users/groups in Makefile and warn if they don't
	exist
o	documented the users and groups necessary for the testcases to run
o	simplified telnet01 when looking for root prompt
o	removed incorrect testcase from sendfile03
o	fixed modify_ldt01 test problems on newer kernels
o	added setrlimit to unlimit core size in waitpid05 test to work 
	around systems where this is set to 0
o	fixed mmstress pthread hang problem

LTP-20010925
------------
o	Testcases have been organized into categories in a directory 
	structure under the testcases subdirectory
o	About 400 new testcases added covering various system calls
o	Andreas Jaeger got confirmation from the glibc people that the
	ulimit test was broken as a result of some changes made in 
	glibc2.2.5.  Since it's the test that's broken, Nate Straz 
	removed the offending testcase.
o	Updated README and INSTALL files
o	Descriptions of the testcases have been added to the doc 
	directory and are organized by type (syscalls.txt, fs.txt, etc.)

------------
2000-09-18 Juan J. Quintela <quintela@fi.udc.es>
           Aaron Laffin <alaffin@sgi.com>

  * tests/mmap001.c: first checkin of a modified version of
  mmap001.c from the memtest suite.

2000-09-08  Egor Duda <deo@logos-m.ru>

	* tests/fpathconf01.c: Close opened file at exit.
	* tests/select01.c: Ditto.

2000-09-08  Andrea Arcangeli <andrea@suse.de>
  * tests/fcntl09.c, tests/fcntl10.c: Rearrange the F_SETLK test to
  to correctly use F_RDLCK and F_WRLCK.

2000-09-06  Egor Duda  <deo@logos-m.ru>

	* lib/parse_opts.c (parse_opts): Initialize allocated string
	to prevent heap corruption.



testcases/fivextra/ps/ps.sh
#!/bin/bash
################################################################################
##                                                                            ##
## (C) Copyright IBM Corp. 2003						      ##
##                                                                            ##
## This program is free software;  you can redistribute it and#or modify      ##
## it under the terms of the GNU General Public License as published by       ##
## the Free Software Foundation; either version 2 of the License, or          ##
## (at your option) any later version.                                        ##
##                                                                            ##
## This program is distributed in the hope that it will be useful, but        ##
## WITHOUT ANY WARRANTY; without even the implied warranty of MEECHANTABILITY ##
## or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License   ##
## for more details.                                                          ##
##                                                                            ##
## You should have received a copy of the GNU General Public License          ##
## along with this program;  if not, write to the Free Software               ##
## Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA    ##
##                                                                            ##
################################################################################
#
# File :        ps.sh
#
# Description:  This testcase tests the commands in the ps package. 
#	free 	killall lsdev 	pgrep 	 pkill 	procinfo 
#	pstree 	skill 	snice 	socklist tload 	top 
#	uptime 	vmstat 	w watch fuser ps sysctl
#              
# 
# Author:       Andrew Pham, apham@us.ibm.com
#
# History:      Feb 06 2003 - Created - Andrew Pham.
#  		Apr 17 2003 - fixed bug in testcases pstree -p and top -bnp.
#			Also changed all testcases to use tc_utils.source.
#			Andrew Pham
#		Apr 22 2003 - Fixed a typo: passfaill -> tc_pass_or_fail
#			Andrew Pham
#		Jul 29 2003 - Fixed sysctl tc to reset the hostname.
#			Andrew Pham
#		Oct 09 2003 - Fixed timming problem.
#		Oct 22 2003 - Fixed killall -e tc problem
#			    - Fixed unnecessary debug output in calculating
#				TST_TOTAL
#		07 Jan 2004 - (apham) updated to tc_utils.source
#		23 Jan 2004 (rcp) swapped test24 and test24xx since the
#				real test now works.
#		11 Feb 2004 (rcp) rewrote fuser test. BUG 6307
#		07 May 2004 (rcp) Rewrote fuser test again.
#				Added installation test.
#				Refuse to run against busybox version of ps.
#				Cleanup TST_TOTAL calculations.

# source the utility functions
me=`which $0`
LTPBIN=${me%%/testcases/*}/testcases/bin
source $LTPBIN/tc_utils.source

# required supporting utilities
REQUIRED="ps grep cat chmod sleep wc kill" 

# commands in the ps package
commands=" ps free killall lsdev pgrep pkill procinfo \
         pstree skill snice socklist top uptime vmstat w ps sysctl"

################################################################################
# global variables
################################################################################

looper=""
myscript=""
myfile=""

# Initialize output messages
summary2="Not available."
summary3="Unexpected output."
################################################################################
# utility functions
################################################################################

#
# be sure none of my sleep commands are running
#
function no_sleep()
{
	ps | grep sleep |
	while read pid garbage
	do
		kill -9 $pid &>/dev/null
	done
}

#
#	tc_local_cleanup
#
function tc_local_cleanup()
{
	[ "$looper" ] && killall $looper &>/dev/null
	no_sleep
}

################################################################################
# the testcase functions
################################################################################

function test00()
{
	tc_register "installation check"
	tc_executes $commands
	tc_pass_or_fail $? "ps package not installed properly"
}

function test1()
{
	tc_register "ps"
	
	ps >$stdout 2>$stderr 
	tc_fail_if_bad $?  "$summary2" || return
	
	no_sleep
	
	sleep 22m &
	sleep 1
	ps > $TCTMP/res001
	grep -q sleep $TCTMP/res001
	tc_pass_or_fail $?  "$summary3"
}

function test2()
{
	tc_register "ps -e"

	ps > $TCTMP/tmp_ps
	ps -e > $TCTMP/tmp_pse 2>$stderr
	tc_fail_if_bad $?  "$summary2" || return
	
	#check to see that ps -e output more stuffs than ps
	set `wc -l $TCTMP/tmp_pse`
	pse_l=$1
	set `wc -l $TCTMP/tmp_ps`
	ps_l=$1

	[ $pse_l -gt $ps_l ]
	tc_pass_or_fail $? " Failed: Unexpected output ."
}

function test3()
{
	tc_register "ps -f"
	
	ps > $TCTMP/tmp_ps 2>/dev/null
	ps -f > $TCTMP/tmp_psf 2>$stderr
	tc_fail_if_bad $?  "$summary2" || return

	#check to see that ps -f output more stuffs than ps
	set `wc -w $TCTMP/tmp_psf`
	psf_w=$1
	set `wc -w $TCTMP/tmp_ps`
	ps_w=$1

	[ $psf_w -gt $ps_w ]
	tc_pass_or_fail $? "$summary3"
}

function test4()
{
	tc_register "free"
	local RC=0

	free >$stdout 2>$stderr
	tc_fail_if_bad $?  "$summary2" || return
	
	for O in V b k m o
	do
		free -$O >$stdout 2>$stderr
		tc_fail_if_bad $?  "free -$O : Not work." || return
		if [ ! -s $stdout ]; then
			RC=1
			break
		fi
	done

	[ "$RC" -eq 0 ]
	tc_pass_or_fail $? "free -$O : $summary3"
}

function test5()
{
	tc_register "free -t"

	free -t >$stdout 2>$stderr 
	tc_fail_if_bad $?  "$summary2" || return

	grep -q Total $stdout
	tc_pass_or_fail $?  "$summary3"
}

function test6()
{
	tc_register "killall -l"
	
	killall -l >$stdout 2>$stderr 
	tc_fail_if_bad $?  "$summary2" || return

	grep -q HUP $stdout
	tc_pass_or_fail $?  "$summary3"
}

function test9()
{
	tc_register "killall -e"

	sleep 234m &
	local tmpPID=`pgrep -n sleep`
	
	sleep 2

	killall -e sleep >$stdout 2>$stderr
	tc_fail_if_bad $?  "$summary2" || return 

	sleep 2
	ps --pid $tmpPID >&/dev/null
	[ ! $? -eq 0 ]
	tc_pass_or_fail $? "$summary3: NOT expect to see proc $tmpPID: `ps --pid $tmpPID`" 
}

function test8()
{
	tc_register "lsdev"

	lsdev  >$stdout 2>$stderr
	tc_fail_if_bad $?  "$summary2" || return

	[ -s $stdout ]
	tc_pass_or_fail $?  "$summary3"
}

function test7()
{
	tc_register "pgrep -n"

	sleep 50m &
	pgrep -n >$stdout 2>$stderr
	tc_fail_if_bad $?  " $summary2" || return
	sleep 2

	local tmpPID=`cat $stdout`

	ps -ef | grep -q $tmpPID
	tc_pass_or_fail $?  "$summary3" || return
	
	tc_register "pgrep -l"
	let TST_TOTAL+=1
	pgrep -l sleep >$stdout 2>$stderr
	tc_fail_if_bad $?  " $summary2" || return

	grep -q sleep $stdout
	tc_pass_or_fail $?  "$summary3" || return

	if [ -n "$UID" ]; then
		tc_register "pgrep -U"
		let TST_TOTAL+=1
		pgrep -U $UID sleep >$stdout 2>$stderr
		tc_fail_if_bad $?  " $summary2" || return
		
		#cat $stdout | grep $tmpPID  # $stdout
		grep -q $tmpPID  $stdout
		tc_pass_or_fail $?  "$summary3"
	fi
}


function test10()
{
	tc_register "pkill"

	sleep 234m &
	local tmpPID=`pgrep -n sleep`
	sleep 1
	pkill sleep >$stdout 2>$stderr 
	tc_fail_if_bad $?  "$summary2" || return
	sleep 2

	ps -ef > $TCTMP/ps_out
	grep -q $tmpPID $TCTMP/ps_out
	[ $? -ne 0 ]
	tc_pass_or_fail $? "$summary3" 
}
function test11()
{
	tc_register "procinfo"

	procinfo -a >$stdout 2>$stderr
	tc_fail_if_bad $? "$summary2" || return

	[ -s $stdout ]
	tc_pass_or_fail $? "$summary3" 
}

function test12()
{
	tc_register "pstree"

	pstree >$TCTMP/pst_tmp 2>$stderr
	tc_fail_if_bad $?  "$summary2" || return
  		
	[ -s $TCTMP/pst_tmp ]
	tc_pass_or_fail $? "$summary3"
}

function test13()
{
	tc_register "pstree -a"

	pstree -a >$TCTMP/pst_tmp_a
	tc_fail_if_bad $?  "$summary2" || return

	#check to see that pstree -a outputs more stuffs than pstree
	set `wc -l $TCTMP/pst_tmp_a`
	pst_a=$1
	set `wc -l $TCTMP/pst_tmp`
	pst=$1

	[ $pst_a -gt $pst ]
	tc_pass_or_fail $? "$summary3"
}

function test14()
{
	tc_register "pstree -p"

	pstree -p > $TCTMP/ps_p.tst 2>$stderr
	tc_fail_if_bad $?  "$summary2" || return

	[ -s $TCTMP/ps_p.tst ]	
	tc_pass_or_fail $? "$summary3"
}

function test15()
{
	tc_register "skill -v"

	sleep 234m &
	sleep 2
	skill -v sleep >$stdout 2>$stderr
	tc_fail_if_bad $?  "$summary2" || return
	
	[ -s $stdout ]
	tc_pass_or_fail $? "$summary3" || return

	# test skill -KILL
	tc_register "skill -KILL"
	let TST_TOTAL+=1

	sleep 235m &
	sleep 2
	skill  -KILL sleep >$stdout 2>$stderr
	tc_fail_if_bad $?  "$summary2" || return 

	sleep 2

	ps -ef > $TCTMP/ps_out
	grep -q sleep $TCTMP/ps_out
	[ $? -ne 0 ]
	tc_pass_or_fail $? "$summary3"
}

function test16()
{
	tc_register "snice -v"

	sleep 234m &
	sleep 2

	# wait a bit for above process to kick in.
	local pid=`pgrep -n`
	local CNT=0

	snice -v  sleep >$stdout 2>$stderr 
	tc_fail_if_bad $?  " $summary2" || return

	[ -s $stdout ]
	tc_pass_or_fail $? " $summary3" || return

	tc_register "snice +17"
	let TST_TOTAL+=1

	snice +17 sleep >$stdout 2>$stderr 
	tc_pass_or_fail $?  "$summary2" || return

	# kill all sleep commands runing
	no_sleep	
	return 0
}
function test17()
{
	tc_register "socklist" 

	socklist >$stdout 2>$stderr 
	tc_fail_if_bad $?  " $summary2" || return

	grep -q inode $stdout
	tc_pass_or_fail $?  " $summary3"
}

function test18()
{
	tc_register "tload"
	tc_info "tload must be tested manually."
}

function test19()
{
	tc_register "top -bn"
	
	top -b -n 1 >$stdout 2>$stderr 
	tc_fail_if_bad $?  " $summary2" || return

	grep -q "1.*root.*init" $stdout
	tc_pass_or_fail $?  " $summary2"
}

function test20()
{
	tc_register "top -bnp"

	sleep 123m & >/dev/null
	local tmpPID=`pgrep -n sleep`
	sleep 2

	top -b -n 1  -p $tmpPID >$stdout 2>$stderr 
	tc_fail_if_bad $?  " $summary2" || return 

	grep -q $tmpPID $stdout
	tc_pass_or_fail $? " $summary3"
}

function test21()
{
	tc_register "uptime"

	uptime >$stdout 2>$stderr 
	tc_fail_if_bad $?  " $summary2" || return

	grep -q up $stdout
	tc_pass_or_fail $?  " $summary3"
}

function test22()
{
	tc_register "vmstat"

	sleep 234m &
	sleep 2
	vmstat >$stdout 2>$stderr 
	tc_fail_if_bad $?  " $summary2" || return

	grep -q cpu $stdout
	tc_pass_or_fail $?  " $summary3" || return

	tc_register "vmstat 1 5"
	let TST_TOTAL+=1

	vmstat 1 5 >$stdout 2>$stderr 
	tc_fail_if_bad $?  "$summary2" || return 

	set `wc -l $stdout`
	[ $1 -eq 7 ]
	tc_pass_or_fail $? "$summary3"
}

function test23()
{
	tc_register "w"

	w >$stdout 2>$stderr 
	tc_fail_if_bad $?  " $summary2" || return 

	grep -q JCPU $stdout
	tc_pass_or_fail $?  " $summary3" || return 

	tc_register "w -s"
	let TST_TOTAL+=1

	w -s >$stdout 2>$stderr 
	tc_fail_if_bad $?  " -s $summary2" || return 

	grep -q JCPU $stdout
	[ $? -ne 0 ]
	tc_pass_or_fail $? "$summary3"
}

function test24()
{
	tc_register "fuser"

	# create script that will loop forever
	# then redirect its output (of which there is none) to a file
	looper=looper$$.sh
	myscript=$TCTMP/$looper
	myfile=$TCTMP/myfile
	cat > $myscript <<-EOF
		#!$SHELL
		while : ; do
			sleep 1
		done
	EOF
	chmod +x $myscript
	$myscript > $myfile &		# will run until fuser -k
	local pid=$!			# pid of $myscript

	# wait for process to start
	local i=0;
	while [ $i -lt 10 ] ; do
		let i+=1
		ps -ef | grep -q "[l]ooper$$.sh" && break
		tc_info "waiting for $looper to start"
		sleep 1
	done
	ps -ef | grep -q "[l]ooper$$.sh"
	tc_break_if_bad $? "could not start a process for fuser -k to kill" || return

	# fuser should show pid for the file
	fuser $myfile >$stdout 2>$stderr
	tc_fail_if_bad $? "unexpected response from fuser" || return
	grep -q "$pid" $stdout 2>$stderr
	tc_fail_if_bad $? "fuser didn't show PID $pid" || return

	tc_info "OK to see killed message for pid $pid"
	# should kill the $pid process
	fuser -k $myfile >$stdout 2>$stderr
	tc_fail_if_bad $? "Unexpected response from fuser -k"

	# wait for process to be killed
	local i=0;
	while [ $i -lt 10 ] ; do
		let i+=1
		ps -ef | grep -v grep | grep "$pid" >$stdout 2>$stderr
		[ -s $stdout ] || break
	done
	[ ! -s $stdout ]
	tc_pass_or_fail $? "process $pid was not killed by fuser -k"
	
}

function test25()
{
	tc_register "sysctl -a"
	local NEW="testing$$"
	local OLD=`hostname`

	# make sure we have root privilege
	tc_root_or_break || return

	sysctl -a >$stdout 2>$stderr
	tc_fail_if_bad $?  "$summary2" || return

	grep -q 'kernel.hostname' $stdout
	tc_pass_or_fail $?  "$summary3" || return

	tc_register "sysctl -w"
	let TST_TOTAL+=1

	sysctl -w kernel.hostname=$NEW >$stdout 2>$stderr 
	tc_fail_if_bad $?  "$summary2" || return

	grep -q $NEW $stdout
	tc_pass_or_fail $?  "$summary3" || return
	
	# resetting hostname
	sysctl -w kernel.hostname=$OLD >& /dev/null
	return 
}

function test26()
{
	tc_register "watch" 

	tc_info "watch must be tested manually."
}
################################################################################
# main
################################################################################
TST_TOTAL=27
ORIG_TST_TOTAL=$TST_TOTAL
tc_setup
tc_exec_or_break $REQUIRED || exit

# ensure not busybox version
! tc_is_busybox ps
tc_break_if_bad $? "Not supported for busybox version" || exit

test00 || exit	# installation check
i=1
while [ $i -lt $ORIG_TST_TOTAL ]
do
	test$i
	let i+=1
done
